lib/invoiced.rb

```diff
@@ -95,7 +109,25 @@ def request(method, endpoint, params={})
             parse(response)
         end
 
-        private
+        def locked_api_request(method, endpoint, params)
+          lock = redis_lock
+          lock.lock
+            api_request(method, endpoint, params)
+          lock.unlock
+        rescue Redis::Lock::AcquireLockTimeOut
+          raise UnableToAcquireLock, "Acquire Lock timed out: #{lock.inspect}"
+        end
+
+        def redis_lock
+          return unless @redis
+
+          Redis::Lock.new(
+            @redis,
+            "invoiced_lock_#{rand(1..ConnectionLimit)}",
+            auto_release_time: 7,
+            base_sleep: 10
```

Comment:
In future we should pull these out into a configuration (along with the connection limit)